### ISSUE TITLE
FTBFS with openJDK9

### ISSUE THREAD

**Issue by hansjoachim** [unclebob/fitnesse#746]
I've tried to test a bit with one of the early openJDK9 builds on Ubuntu Wily and discovered that FitNesse fails to build with JDK9. When running `ant compile`, I get the following error message:
```
    [javac] /home/ubuntu/fitnesse/src/fitnesse/slim/converters/ConverterRegistry.java:48: error: incompatible types: Converter<CAP#1> cannot be converted to Converter<T>
    [javac]     Converter<T> converter = getConverterForClass(clazz, null);
    [javac]                                                  ^
    [javac]   where T is a type-variable:
    [javac]     T extends Object declared in method <T>getConverterForClass(Class<? extends T>)
    [javac]   where CAP#1 is a fresh type-variable:
    [javac]     CAP#1 extends T from capture of ? extends T
``` 

As mentioned this is an early build (specifically 9~b64-1ubuntu1), so the bug is not necessarily in FitNesse, but I thought I'd report it just in case.

**Comment by hansjoachim** [comment]
The lines in question was recently refactored/simplified, but FWIW I still get the same error message
```
    [javac] /home/ubuntu/dev/fitnesse/src/fitnesse/slim/converters/ConverterRegistry.java:53: error: incompatible types: Converter<CAP#1> cannot be converted to Converter<T>
    [javac]     return getConverterForClass(clazz, null);
    [javac]                                ^
    [javac]   where T is a type-variable:
    [javac]     T extends Object declared in method <T>getConverterForClass(Class<? extends T>)
    [javac]   where CAP#1 is a fresh type-variable:
    [javac]     CAP#1 extends T from capture of ? extends T
```
(OpenJDK9 9~b80-2 on Ubuntu Wily)

I have to admit I'm not quite certain of what the error message is trying to say...

**Comment by amolenaar** [comment]
Does this have to do with T being a generic type itself?

**Comment by hansjoachim** [comment]
Might be.

I read a bit on http://stackoverflow.com/questions/25914289/incompatible-types-required-classt-found-classcap1-where-t-is-a-type-vari and have a simple fix/workaround now. For now, I'm just casting the value to Converter<T> before returning which might not be the best, but I can look a bit more into it. With that in place, all the source code builds successfully with JDK9, though unfortunately I ran into a couple of test failures:
```
    [junit] Tests run: 12, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 0.097 sec
    [junit] 
    [junit] Testcase: shouldGetVersionForClassInJarWithVersion took 0.004 sec
    [junit] 	FAILED
    [junit] Version not retrieved for java.lang.reflect.Method
    [junit] junit.framework.AssertionFailedError: Version not retrieved for java.lang.reflect.Method
    [junit] 	at fitnesse.slim.StackTraceEnricherTest.shouldGetVersionForClassInJarWithVersion(StackTraceEnricherTest.java:133)
    [junit] 
    [junit] Testcase: shouldParseDirectories took 0.001 sec
    [junit] Testcase: shouldParseCauseExceptions took 0.001 sec
    [junit] Testcase: shouldGetVersionForStackTraceElementInJarWithVersion took 0.001 sec
    [junit] 	FAILED
    [junit] Version not retrieved for java.lang.reflect.Method
    [junit] junit.framework.AssertionFailedError: Version not retrieved for java.lang.reflect.Method
    [junit] 	at fitnesse.slim.StackTraceEnricherTest.shouldGetVersionForStackTraceElementInJarWithVersion(StackTraceEnricherTest.java:140)
    [junit] 
    [junit] Testcase: shouldWriteToSystemErrByDefault took 0.001 sec
    [junit] Testcase: shouldAddVersionWhenAvailable took 0.001 sec
    [junit] 	FAILED
    [junit] Version not added for rt.jar
    [junit] junit.framework.AssertionFailedError: Version not added for rt.jar
    [junit] 	at fitnesse.slim.StackTraceEnricherTest.shouldAddVersionWhenAvailable(StackTraceEnricherTest.java:127)
    [junit] 
```

Not really sure why they are failing, and I haven't had the chance to debug in an IDE because I haven't found one which supports/runs with JDK9 yet. I haven't tested the proposed fix with older JDKs either, which I also would like to do before suggesting a PR.

**Comment by hansjoachim** [comment]
A bit tedious to debug without an IDE, but I've found the following:
The test first looks up javaVersion from the System property "java.runtime.version", which is set to "1.9.0-internal-b88". It then tries to compare this against various version numbers it tries to read out via the StackTraceEnricher. 

I've looked mostly at shouldGetVersionForClassInJarWithVersion (java.lang.reflect.Method.class) which is unable to find the version and ends up with the default "n/a". This lookup happens in ClassMetaInformation::getVersion (https://github.com/unclebob/fitnesse/blob/master/src/fitnesse/slim/StackTraceEnricher.java#L189). Based on printlines it finds the package, but doesn't match either of the ifs so it ends up in the non-existing else case:
```
            if (pack.getImplementationVersion() != null) {
              version = pack.getImplementationVersion();
            } else if (pack.getSpecificationVersion() != null) {
              version = pack.getSpecificationVersion();
            }
```
Thus the method ends up returning the default version. Possibly this section needs to be expanded with another else if in order to be able to find the version number, though I am not sure how.

Normally (at least with JDK8), this test finds the value of pack.getImplementationVersion and returns that. I note that this might return null if the version is not known (http://docs.oracle.com/javase/8/docs/api/java/lang/Package.html#getImplementationVersion--), and it might be affected by running milestone builds(?). I looked up the Java9 javadoc locally and it doesn't seem to have any additional comments on this method either...
